test/lsp.py

```diff
@@ -322,17 +742,72 @@ def wait_for_diagnostics(self, solc: JsonRpcProcess, count: int) -> List[dict]:
         Return `count` number of published diagnostic reports sorted by file URI.
         """
         reports = []
+
+#        import pdb; pdb.set_trace()
         for _ in range(0, count):
-            message = solc.receive_message()
+            try:
+                message = solc.read_message(peek=True, block=True)
+            except queue.Empty:
+                break
+
             assert message is not None # This can happen if the server aborts early.
-            reports.append(
-                self.require_params_for_method(
-                    'textDocument/publishDiagnostics',
-                    message,
+
+            if "method" in message and message["method"] == "textDocument/publishDiagnostics":
+                reports.append(
+                    self.require_params_for_method(
+                        'textDocument/publishDiagnostics',
+                        message,
+                    )
                 )
-            )
+                solc.read_message(peek=False) # remove peeked message
+            else:
+                break
+
         return sorted(reports, key=lambda x: x['uri'])
 
+    def fetch_and_format_diagnostics(self, solc: JsonRpcProcess, test):
+        expectations = ""
+
+        published_diagnostics = self.open_file_and_wait_for_diagnostics(solc, test, 100)
+
+        for diagnostics in published_diagnostics:
+            testname = diagnostics["uri"].replace(self.project_root_uri + "/", "")[:-len(".sol")]
+
+            # Skip empty diagnostics within the same file
+            if len(diagnostics["diagnostics"]) == 0 and testname == test:
+                continue
+
+            expectations += f"// {testname}:"
+
+            for diagnostic in diagnostics["diagnostics"]:
+                tag = self.find_tag_with_range(testname, diagnostic['range'])
+
+                if tag is None:
+                    raise Exception(f"No tag found for diagnostic range {diagnostic['range']}")
+
+                expectations += f" {tag} {diagnostic['code']}"
+            expectations += "\n"
+
+        return expectations
+
+    def update_diagnostics_in_file(
+        self,
+        solc: JsonRpcProcess,
+        test,
+        content,
+        current_diagnostics:
+        TestParser.Diagnostics
+    ):
+        content = content[:current_diagnostics.start] + \
```

Comment:
maybe use real if statements (not expression) here to make the construction more readable?